src/git-repository-async.js

```diff
@@ -0,0 +1,870 @@
+'use babel'
+
+import fs from 'fs-plus'
+import Git from 'nodegit'
+import path from 'path'
+import {Emitter, CompositeDisposable, Disposable} from 'event-kit'
+
+const modifiedStatusFlags = Git.Status.STATUS.WT_MODIFIED | Git.Status.STATUS.INDEX_MODIFIED | Git.Status.STATUS.WT_DELETED | Git.Status.STATUS.INDEX_DELETED | Git.Status.STATUS.WT_TYPECHANGE | Git.Status.STATUS.INDEX_TYPECHANGE
+const newStatusFlags = Git.Status.STATUS.WT_NEW | Git.Status.STATUS.INDEX_NEW
+const deletedStatusFlags = Git.Status.STATUS.WT_DELETED | Git.Status.STATUS.INDEX_DELETED
+const indexStatusFlags = Git.Status.STATUS.INDEX_NEW | Git.Status.STATUS.INDEX_MODIFIED | Git.Status.STATUS.INDEX_DELETED | Git.Status.STATUS.INDEX_RENAMED | Git.Status.STATUS.INDEX_TYPECHANGE
+
+// Just using this for _.isEqual and _.object, we should impl our own here
+import _ from 'underscore-plus'
+
+export default class GitRepositoryAsync {
+  static open (path, options = {}) {
+    // QUESTION: Should this wrap Git.Repository and reject with a nicer message?
+    return new GitRepositoryAsync(path, options)
+  }
+
+  static get Git () {
+    return Git
+  }
+
+  constructor (_path, options) {
+    this.repo = null
+    this.emitter = new Emitter()
+    this.subscriptions = new CompositeDisposable()
+    this.pathStatusCache = {}
+    this.repoPromise = Git.Repository.open(_path)
+    this.isCaseInsensitive = fs.isCaseInsensitive()
+    this.upstreamByPath = {}
+
+    this._refreshingCount = 0
+
+    options = options || {}
+
+    let {refreshOnWindowFocus = true} = options
+    if (refreshOnWindowFocus) {
+      const onWindowFocus = () => this.refreshStatus()
+      window.addEventListener('focus', onWindowFocus)
+      this.subscriptions.add(new Disposable(() => window.removeEventListener('focus', onWindowFocus)))
+    }
+
+    const {project, subscribeToBuffers} = options
+    this.project = project
+    if (this.project && subscribeToBuffers) {
+      this.project.getBuffers().forEach(buffer => this.subscribeToBuffer(buffer))
+      this.subscriptions.add(this.project.onDidAddBuffer(buffer => this.subscribeToBuffer(buffer)))
+    }
+  }
+
+  // Public: Destroy this {GitRepositoryAsync} object.
```

Comment:
I don't think so. I think that's what https://github.com/atom/donna/issues/9 is about.